spec.md

```diff
@@ -2224,15 +2246,28 @@ If the volume corresponding to the `volume_id` has already been published at the
 
 If this RPC failed, or the CO does not know if it failed or not, it MAY choose to call `NodePublishVolume` again, or choose to call `NodeUnpublishVolume`.
 
-This RPC MAY be called by the CO multiple times on the same node for the same volume with possibly different `target_path` and/or other arguments if the volume has MULTI_NODE capability (i.e., `access_mode` is either `MULTI_NODE_READER_ONLY`, `MULTI_NODE_SINGLE_WRITER` or `MULTI_NODE_MULTI_WRITER`).
+This RPC MAY be called by the CO multiple times on the same node for the same volume with possibly different `target_path` and/or other arguments if the volume has MULTI_NODE or SINGLE_NODE_MULTI_WRITER (see second table) access modes.
+Example MULTI_NODE access modes include `MULTI_NODE_READER_ONLY`, `MULTI_NODE_SINGLE_WRITER` or `MULTI_NODE_MULTI_WRITER`.
 The following table shows what the Plugin SHOULD return when receiving a second `NodePublishVolume` on the same volume on the same node:
 
+(`Tn`: target path of the n-th `NodePublishVolume`, `Pn`: other arguments of the n-th `NodePublishVolume` except `secrets`)
+
 |                | T1=T2, P1=P2    | T1=T2, P1!=P2  | T1!=T2, P1=P2       | T1!=T2, P1!=P2     |
 |----------------|-----------------|----------------|---------------------|--------------------|
 | MULTI_NODE     | OK (idempotent) | ALREADY_EXISTS | OK                  | OK                 |
 | Non MULTI_NODE | OK (idempotent) | ALREADY_EXISTS | FAILED_PRECONDITION | FAILED_PRECONDITION|
 
-(`Tn`: target path of the n-th `NodePublishVolume`, `Pn`: other arguments of the n-th `NodePublishVolume` except `secrets`)
+NOTE: If the Plugin supports the `SINGLE_NODE_MULTI_WRITER` capability, use the following table instead for what the Plugin SHOULD return when receiving a second `NodePublishVolume` on the same volume on the same node:
+
+|                                       | T1=T2, P1=P2    | T1=T2, P1!=P2  | T1!=T2, P1=P2       | T1!=T2, P1!=P2      |
+|---------------------------------------|-----------------|----------------|---------------------|---------------------|
+| SINGLE_NODE_SINGLE_WRITER             | OK (idempotent) | ALREADY_EXISTS | FAILED_PRECONDITION | FAILED_PRECONDITION |
+| SINGLE_NODE_MULTI_WRITER              | OK (idempotent) | ALREADY_EXISTS | OK                  | OK                  |
```

Comment:
For many volume types - I suspect they may not support consuming same volume with different capability on same node. For  example - would they allow mounting volume with different mount flags or would they allow consuming volume both as block and fs volume at same time?

So I am bit doubtful that if "P1!=P2" and "T1!=T2" - node-publish will be okay (assuming driver really enforces supplied capabilities).  I suspect this will lead to drivers ignoring certain capabilities and allowing node-publish anyways. It may be okay, but strictly speaking they are not spec compliant then.

Reply:
We currently make the clarification that `Pn` refers to other arguments passed to `NodePublishVolume` except `secrets`. We could extend this to also include `volume_capability`. If we want to play it safe we could make this change just for the new table. Does this sound reasonable?

Reply:
I think the *except `secrets`* wording does not make it clear, if calling `NodePublish` multiple times on same node has to be done with same secret or secrets can be different. I always found that wording to be ambiguous, so when we add `volume_capability` to it - does it mean the capabilities has to be same?  If yes - may be we should add more fields to the table. Something like:

```
T1!=T2, P1!=P2, Cap1!=Cap2 -> FAILED_PRECONDITION
T1!=T2, P1!=P2, Cap1==Cap2 -> OK
```

Reply:
Discussed this in the CSI standup, we're adding a sentence to make this explicit to avoid adding another column and more rows to the table. Added this sentence:
> COs SHOULD NOT call `NodePublishVolume` a second time with different `volume_capability`. If this happens, the Plugin SHOULD return `FAILED_PRECONDITION`.

Reply:
@bswartz for review on this

Reply:
+1 I like it

Reply:
Thanks. I think having explicit wording about capability helps. +1